modules/ml/src/boost.cpp

```diff
@@ -490,6 +490,7 @@ class BoostImpl : public Boost
 
     float predict( InputArray samples, OutputArray results, int flags ) const CV_OVERRIDE
     {
+        CV_Assert( samples.cols() == getVarCount() && samples.type() == CV_32F );
```

Comment:
@danielenricocahall would you mind explain why this `assert` was in here?
The test from `text` module was relying on this code, and it was passing before this PR, but now it raises this assertion.

### Before
```
[==========] Running 12 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 12 tests from Text/Detection
[ RUN      ] Text/Detection.sample/0, where GetParam() = ("text/scenetext01.jpg", false)
Image: text/scenetext01.jpg
Orientation: horiz
Found groups: 25
[       OK ] Text/Detection.sample/0 (4981 ms)
[ RUN      ] Text/Detection.sample/1, where GetParam() = ("text/scenetext01.jpg", true)
[     SKIP ] ERGROUPING_ORIENTATION_ANY mode is not supported
[       OK ] Text/Detection.sample/1 (4 ms)
  :
[==========] 12 tests from 1 test case ran. (37679 ms total)
[  PASSED  ] 12 tests.
```

### After
```
[==========] Running 12 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 12 tests from Text/Detection
[ RUN      ] Text/Detection.sample/0, where GetParam() = ("text/scenetext01.jpg", false)
Image: text/scenetext01.jpg
Orientation: horiz
unknown file: Failure
C++ exception with description "OpenCV(3.4.14-dev) /opencv-fork/modules/ml/src/boost.cpp:493: error: (-215:Assertion failed) samples.cols() == getVarCount() && samples.type() == CV_32F in function 'predict'
" thrown in the test body.
[  FAILED  ] Text/Detection.sample/0, where GetParam() = ("text/scenetext01.jpg", false) (68 ms)
[ RUN      ] Text/Detection.sample/1, where GetParam() = ("text/scenetext01.jpg", true)
[     SKIP ] ERGROUPING_ORIENTATION_ANY mode is not supported
[       OK ] Text/Detection.sample/1 (4 ms)
  :
[==========] 12 tests from 1 test case ran. (861 ms total)
[  PASSED  ] 6 tests.
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] Text/Detection.sample/0, where GetParam() = ("text/scenetext01.jpg", false)
[  FAILED  ] Text/Detection.sample/2, where GetParam() = ("text/scenetext02.jpg", false)
[  FAILED  ] Text/Detection.sample/4, where GetParam() = ("text/scenetext03.jpg", false)
[  FAILED  ] Text/Detection.sample/6, where GetParam() = ("text/scenetext04.jpg", false)
[  FAILED  ] Text/Detection.sample/8, where GetParam() = ("text/scenetext05.jpg", false)
[  FAILED  ] Text/Detection.sample/10, where GetParam() = ("text/scenetext06.jpg", false)
```

I'm trying on Jetson series and Raspberry Pis but all the test results are pointing that this assertion is too strict.
May be `samples.type() == CV_32F` could be removed?

Reply:
I was just using the same assertion that is used in other `predict` methods:
- https://github.com/opencv/opencv/blob/master/modules/ml/src/svm.cpp#L2013
- https://github.com/opencv/opencv/blob/master/modules/ml/src/knearest.cpp#L370
- https://github.com/opencv/opencv/blob/master/modules/ml/src/nbayes.cpp#L314
would it make sense to remove it, or is there something in the `text` test that should be updated?

Reply:
> May be samples.type() == CV_32F could be removed?

In general it should be removed. Types should be defined by underlying classifiers in Boost (e.g, see `DTreesImpl::predictTrees`).

Unfortunately, in opencv_text case error is raised on the first part of condition:

```
C++ exception with description "OpenCV(3.4.14-dev) /home/alalek/projects/opencv/dev/modules/ml/src/boost.cpp:493: error: (-2:Unspecified error) in function 'virtual float cv::ml::BoostImpl::predict(cv::InputArray, cv::OutputArray, int) const'
>  (expected: 'samples.cols() == getVarCount()'), where
>     'samples.cols()' is 4
> must be equal to
>     'getVarCount()' is 5
" thrown in the test body.
```

So, opencv_text module should be checked / fixed too.

Reply:
Problem is caused by invalid import in "ml" module (misused "legacy" mode due to bug #5412).
opencv_text code looks valid.